arch/xtensa/core/userspace.S

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2022, Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <xtensa-asm2-s.h>
+#include <offsets.h>
+#include <offsets_short.h>
+#include <zephyr/syscall.h>
+#include <zsr.h>
+
+/**
+ *  syscall number     arg1, arg2, arg3, arg4, arg5, arg6
+ *  --------------     ----------------------------------
+ *  a2                 a6,   a3,   a4,   a5,   a8,   a9
+ *
+ **/
+.pushsection .text.z_xtensa_do_syscall, "ax"
+.global	z_xtensa_do_syscall
+.align	4
+z_xtensa_do_syscall:
+	rsr a0, ZSR_CPU
+	l32i a0, a0, ___cpu_t_current_OFFSET
+	l32i a0, a0, _thread_offset_to_psp
+
+	addi a0, a0, -___xtensa_irq_bsa_t_SIZEOF
+
+	s32i a1, a0, ___xtensa_irq_bsa_t_scratch_OFFSET
+	s32i a2, a0, ___xtensa_irq_bsa_t_a2_OFFSET
+	s32i a3, a0, ___xtensa_irq_bsa_t_a3_OFFSET
+	rsr a2, ZSR_ALLOCA
+	s32i a2, a0, ___xtensa_irq_bsa_t_a0_OFFSET
+	rsr.ps a2
+	s32i a2, a0, ___xtensa_irq_bsa_t_ps_OFFSET
+	rsr.epc1 a2
+	s32i a2, a0, ___xtensa_irq_bsa_t_pc_OFFSET
+
+	movi a2, PS_WOE|PS_INTLEVEL(XCHAL_NMILEVEL)
+	rsr.ps a3
+	or a3, a3, a2
+	movi a2, ~(PS_EXCM | PS_RING(3))
+	and a3, a3, a2
+	wsr.ps a3
+	rsync
+	l32i a2, a0, ___xtensa_irq_bsa_t_a2_OFFSET
+	l32i a3, a0, ___xtensa_irq_bsa_t_a3_OFFSET
+	SPILL_ALL_WINDOWS
```

Comment:
I'm especially confused by the spilling of register windows.  Syscalls, like exceptions, return back into the interrupted context unconditionally.  We want to leave those registers in place and let them spill naturally.

Reply:
This is done before switch to privileged stack and the intention is to avoid mixing user/kernel context. Though, we have already used a0-3 in this window ...